runtime/lua/vim/filetype.lua

```diff
@@ -2283,13 +2298,13 @@ end
 
 ---@private
 local function dispatch(ft, path, bufnr, ...)
+  local extra
   if type(ft) == 'function' then
-    ft = ft(path, bufnr, ...)
+    ft, extra = ft(path, bufnr, ...)
```

Comment:
+1 for making side effects opt-in. This is an unusual pattern, so I wonder if we can make it more intuitive by thinking of a name other than "extra". Perhaps `bufinit`, `init`, `setup`? Any name that would make sense as a normal function.

If this isn't user-facing, then it's less important. But still helpful.

```suggestion
    ft, setup = ft(path, bufnr, ...)
```